### PR TITLE
added model disagreement sampler

### PR DESCRIPTION
**Pull request by musslick**
## Description

This feature implements a model disagreement sampler for both classification and regression models.

The sampler computes the predictions for a given sample of independent variables and then identifies the independent variables that yield the highest disagreement between N models. 

The PR comes with two unit tests, one for a classification model and one for a regression model.

_Fixes #(issue)_ (if the issue is on Github, simply link to it using the '#' symbol; otherwise, provide a notion page link)

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- momdel disagreement sampler
- unit test for classification models
- unit test for the regression models

### CODE REVIEW

**Review by hollandjg — COMMENTED** [review]
This looks great! I suspect we can make the function a bit simpler if we split it into two. What do you think?

File: autora/experimentalist/sampler/model_disagreement.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            # get predictions from both models
+            Y_A = model_A_predict(X_predict)
+            Y_B = model_B_predict(X_predict)
+
+            if Y_A.shape != Y_B.shape:
+                raise Exception("Models must have same output shape.")
+
+            # determine the disagreement between the two models in terms of mean-squared error
+            if len(Y_A.shape) == 1:
+                disagreement = (Y_A - Y_B) ** 2
```
[the remark this answers is not in the document]
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py:57]: Hm but it passes the unit test for the models with `predict_proba` method? Not sure why it wouldn't work?

File: autora/experimentalist/sampler/model_disagreement.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                continue
+
+            # determine the prediction method
+            model_A_predict = getattr(model_A, "predict_proba", None)
+            if callable(model_A_predict) is False:
+                model_A_predict = getattr(model_A, "predict", None)
+
+            model_B_predict = getattr(model_B, "predict_proba", None)
+            if callable(model_B_predict) is False:
+                model_B_predict = getattr(model_B, "predict", None)
```
[the remark this answers is not in the document]
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py:43]: Yes, that makes sense!

**Comment by musslick** [comment]
> This looks great! I suspect we can make the function a bit simpler if we split it into two. What do you think?

In terms of splitting it up, I was wondering what you mean by having a separate sampler. I might argue against splitting for usability reasons since the user would have to know which sampler to use. Why not have a single sampler function that works for both?

File: autora/experimentalist/sampler/model_disagreement.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+from typing import Iterable, List
+
+import numpy as np
+
+
+def model_disagreement_sampler(X: np.array, models: List, num_samples: int = 1):
```
[the remark this answers is not in the document]
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py:6]: See my main comment: I advocate for keeping in the same function for usability reasons. Also, there is no fundamental distinction between the two types of models.

File: autora/experimentalist/sampler/model_disagreement.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    X_predict = np.array(X)
+    if len(X_predict.shape) == 1:
+        X_predict = X_predict.reshape(-1, 1)
+
+    model_disagreement = list()
+
+    # collect diagreements for each model apir
+    for idx_A, model_A in enumerate(models):
+        for idx_B, model_B in enumerate(models):
```
[the remark this answers is not in the document]
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py:30]: Sounds good, what's the advantage?

File: autora/experimentalist/sampler/model_disagreement.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py:43]: Addressed!

File: autora/experimentalist/sampler/model_disagreement.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py]: Addressed!

File: autora/experimentalist/sampler/model_disagreement.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        if callable(model_a_predict) is False:
+            model_a_predict = getattr(model_a, "predict", None)
+            model_b_predict = getattr(model_b, "predict", None)
+        else:
+            model_b_predict = getattr(model_b, "predict_proba", None)
+            if callable(model_b_predict) is False:
+                raise Exception("Models must have the same prediction method.")
+
+        if model_a_predict is None or model_b_predict is None:
+            raise Exception("Model must have `predict` or `predict_proba` method.")
```
Comment (hollandjg) [inline autora/experimentalist/sampler/model_disagreement.py:43]: how about:
```suggestion
        if hasattr(model_a, "predict_proba") and hasattr(model_b, "predict_proba"):
            model_a_predict = model_a.predict_proba
            model_b_predict = model_b.predict_proba
        elif hasattr(model_a, "predict") and hasattr(model_b, "predict"):
            model_a_predict = model_a.predict
            model_b_predict = model_b.predict
        else:
            raise AttributeError("Models must both have `predict_proba` or `predict` method.")
```

File: autora/experimentalist/sampler/model_disagreement.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+        # get predictions from both models
+        y_a = model_a_predict(X_predict)
+        y_b = model_b_predict(X_predict)
+
+        if y_a.shape != y_b.shape:
+            raise Exception("Models must have same output shape.")
```
Comment (hollandjg) [inline autora/experimentalist/sampler/model_disagreement.py:50]: how about
```suggestion
        assert y_a.shape == y_b.shape, "Models must have same output shape."
```

File: autora/experimentalist/sampler/model_disagreement.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        # determine the disagreement between the two models in terms of mean-squared error
+        if len(y_a.shape) == 1:
+            disagreement = (y_a - y_b) ** 2
+        else:
+            disagreement = np.mean((y_a - y_b) ** 2, axis=1)
+
+        model_disagreement.append(disagreement)
+
+    if len(model_disagreement) == 0:
+        raise Exception("No models to compare.")
```
Comment (hollandjg) [inline autora/experimentalist/sampler/model_disagreement.py:61]: how about:
```suggestion
    assert len(model_disagreement) >= 1, "No disagreements to compare."
```

File: autora/experimentalist/sampler/model_disagreement.py (hunk 5 above)
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py]: Excellent suggestion!

File: autora/experimentalist/sampler/model_disagreement.py (hunk 6 above)
Comment (musslick) [reply autora/experimentalist/sampler/model_disagreement.py:47]: Sounds good!

**Review by hollandjg — COMMENTED** [review]
Minor minor things. Looks great.🎄🚀

**Review by hollandjg — APPROVED** [review]

**Comment by musslick** [comment]
Nicolaus present!

**Pull request merged**